prow/cmd/audit-owners/main.go

```diff
@@ -0,0 +1,182 @@
+var repoRoot string
+
+// The name of the GitHub org to use when querying for collaborators
+var githubOrg string
+
+// The name of the GitHub repo to use when querying for collaborators
+var githubRepo string
+
+func init() {
+	flag.StringVar(&repoRoot, "s", "", "The location of the repo on the local filesystem to start inspecting")
+	flag.StringVar(&githubOrg, "o", "", "The name of the GitHub org to use when querying for collaborators")
+	flag.StringVar(&githubRepo, "r", "", "The name of the GitHub repo to use when querying for collaborators")
+}
+
+func main() {
+
+	flag.Parse()
+
+	// We require a github token to make API calls due to rate limiting
+	if os.Getenv("GITHUB_TOKEN") == "" {
+		fmt.Println("Error: Please supply an environment variable named GITHUB_TOKEN with a valid token")
+		os.Exit(1)
+	}
+
+	// Ensure all the flags were passed in
+	if repoRoot == "" {
+		fmt.Println("Error: Please supply a local source code location with the -s flag")
+		os.Exit(1)
+	}
+	if githubOrg == "" {
+		fmt.Println("Error: Please supply a GitHub org with the -o flag")
+		os.Exit(1)
+	}
+	if githubRepo == "" {
+		fmt.Println("Error: Please supply a GitHub repo name with the -r flag")
+		os.Exit(1)
+	}
+
+	// Stores the usernames found in the OWNERS files
+	handles := make(map[string]struct{})
+
+	// Walk the file tree to gather all the GitHub Logins used in all OWNERS files
+	err := filepath.Walk(repoRoot, func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+
+		name := info.Name()
+
+		// Make sure to skip anything in the .git directory. OWNERS files can be
+		// found in there but their format will cause read errors.
+		if name == ".git" {
+			return filepath.SkipDir
+		}
+
+		// Add in handling for OWNERS_ALIASES files
+		if name == "OWNERS" {
+			// found an OWNERS file. Process it.
+			y, e2 := readOwners(path)
+			if e2 != nil {
+				return e2
+			}
+
+			// TODO: Make sure all names are normalized to lowercase
+			for _, v := range y.Approvers {
+				if _, ok := handles[v]; !ok {
+					handles[v] = struct{}{}
+				}
+			}
+
+			for _, v := range y.Reviewers {
+				if _, ok := handles[v]; !ok {
+					handles[v] = struct{}{}
+				}
+			}
+		}
+
+		return nil
+	})
+
+	if err != nil {
+		fmt.Println("Error walking the directory tree:", err)
+		os.Exit(1)
+	}
+
+	// Query the GitHub API for a repo to get the names of all collaborators.
+	// Note, the merge functionality in prow only requires a collaborator with
+	// read only access to a repo but is also in an OWNERS file.
+
+	// Using the test-infra github client because it can handle pagenation
+	// for long lists
+	// TODO: Allow the API endpoint to be configurable
+	gc := github.NewClient(os.Getenv("GITHUB_TOKEN"), "https://api.github.com")
+
+	users, err := gc.ListCollaborators(githubOrg, githubRepo)
+	if err != nil {
+		fmt.Println("Error getting collaborators", err)
+		os.Exit(1)
+	}
+
+	// Look at all the OWNERS and find which are not collaborators
```

Comment:
If you use `sets.String` from `k8s.io/apimachinery/pkg/util/sets` for `handles` and `users` you can replace the bellow code with just `nameList := handles.Difference(users).List()`.